extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/DataSourceInitializedBuildItem.java

```diff
@@ -1,12 +1,65 @@
 package io.quarkus.agroal.deployment;
 
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+
 import io.quarkus.builder.item.SimpleBuildItem;
 
 /**
  * Marker build item indicating the datasource has been fully initialized.
+ * Contains all processed datasource names, including an empty string for the default datasource.
  */
 public final class DataSourceInitializedBuildItem extends SimpleBuildItem {
+    private static final String DEFAULT_DATASOURCE_NAME = "";
+
+    private final Collection<String> dataSourceNames = new HashSet<>();
+
+    /**
+     * Null-safe way to get the datasource names from the given {@link DataSourceInitializedBuildItem}.
+     * 
+     * @param buildItem {@link DataSourceInitializedBuildItem}
+     * @return {@link Collection} of datasource names as {@link String}s
+     */
+    public static final Collection<String> dataSourceNamesOf(DataSourceInitializedBuildItem buildItem) {
+        return (buildItem != null) ? buildItem.getDataSourceNames() : Collections.emptyList();
+    }
+
+    /**
+     * Null-safe way to find out, if the given {@link DataSourceInitializedBuildItem} contains the default datasource.
+     * 
+     * @param buildItem {@link DataSourceInitializedBuildItem}
+     * @return <code>true</code>, if the default datasource is present
+     */
+    public static final boolean isDefaultDataSourcePresent(DataSourceInitializedBuildItem buildItem) {
+        return (buildItem != null) ? buildItem.isDefaultDataSourcePresent() : false;
+    }
+
+    public DataSourceInitializedBuildItem(Collection<String> dataSourceNames) {
+        this.dataSourceNames.addAll(dataSourceNames);
+    }
+
+    public DataSourceInitializedBuildItem withDefaultDatasource() {
+        Collection<String> dataSourceNamesWithDefault = getDataSourceNames();
+        dataSourceNamesWithDefault.add(DEFAULT_DATASOURCE_NAME);
```

Comment:
As mentioned elsewhere, I would keep the default datasource entirely separate.

You have to special case everywhere anyway.